Frameworks/UIKit/NSLayoutManager.mm

```diff
@@ -34,22 +34,38 @@ @implementation NSLayoutManager {
     woc::unique_cf<CTFrameRef> _frame;
 }
 
-// Returns true if any of the characters in line after index have visible glyphs, false otherwise
-// Used to determine if a given index is past the visible part of a line for linebreaking
-static bool __lineHasGlyphsAfterIndex(CTLineRef line, CFIndex index) {
-    for (id runRef in static_cast<NSArray*>(CTLineGetGlyphRuns(line))) {
-        CTRunRef run = static_cast<CTRunRef>(runRef);
-        CFRange runRange = CTRunGetStringRange(run);
-        if (runRange.location <= index && index <= runRange.location + runRange.length) {
-            const CFIndex* indices = CTRunGetStringIndicesPtr(run);
-            if (std::any_of(indices,
-                            indices + CTRunGetGlyphCount(run),
-                            std::bind(std::greater_equal<CFIndex>(), std::placeholders::_1, index))) {
-                return true;
+// Private helper:
+// Given a CTLineRef, returns the index of the character in the original string, that maps to the last visible glyph in the line
+static CFIndex __CTLineGetStringIndexOfLastVisibleGlyph(CTLineRef line) {
+    CFArrayRef runs = CTLineGetGlyphRuns(line);
+
+    CFIndex ret = -1;
+    for (CFIndex i = CFArrayGetCount(runs) - 1; i >= 0; --i) {
+        CTRunRef run = static_cast<CTRunRef>(CFArrayGetValueAtIndex(runs, i));
+        CFIndex glyphCount = CTRunGetGlyphCount(run);
+
+        // Open question: Can runs ever be non-monotonic with respect to the original string?
```

Comment:
I can't imagine RTL languages would store either their glyphs, runs, or characters back-to-front in their respective arrays?
Also, note that this comment is meant to convey a future optimization that could be made, if it turns out that the assumption _is_ safe. I think the current logic should handle RTL languages just fine,

---
In reply to: [92738146](https://github.com/Microsoft/WinObjC/pull/1558#discussion_r92738146) [](ancestors = 92738146)